test: posix: common: nanosleep test fix

hakehuang opened the pull request:
in test_nanosleep_execution, the

actual_ns = k_cyc_to_ns_ceil64((now - then * selection));

is wrong, according to posix driver code

lib/posix/clock.c

line: 156
	if ((flags & TIMER_ABSTIME) == 0) {
		ns += uptime_ns;
	}

so we need change to

actual_ns = k_cyc_to_ns_ceil64((now + then * selection));

also to increase accuracy in comparison as the sleep is using k_sleep, change to k_cycle_get_64

Fixes: #62889

hakehuang committed:
test: posix: common: nanosleep test fix

in test_nanosleep_execution, the

actual_ns = k_cyc_to_ns_ceil64((now - then * selection));

is wrong, according to posix driver code

lib/posix/clock.c

line: 156
	if ((flags & TIMER_ABSTIME) == 0) {
		ns += uptime_ns;
	}

so we need change to

actual_ns = k_cyc_to_ns_ceil64((now + then * selection));

fixing: #62889

Co-authored-by: Christopher Friedt <cfriedt@meta.com>
Signed-off-by: Hake Huang <hake.huang@oss.nxp.com>

hakehuang commented:
tested PASS in RT1060_evk
'''
*** Booting Zephyr OS build v3.5.0-rc2-1-gee588bae7a61 ***
Running TESTSUITE posix_apis
===================================================================
START - test_barrier
 PASS - test_barrier in 0.001 seconds
===================================================================
START - test_clock
POSIX clock APIs
POSIX clock APIs test done
 PASS - test_clock in 1.005 seconds
===================================================================
START - test_clock_nanosleep_errors_errno
 PASS - test_clock_nanosleep_errors_errno in 0.001 seconds
===================================================================
START - test_clock_nanosleep_execution
 PASS - test_clock_nanosleep_execution in 3.942 seconds
===================================================================
START - test_cond_resource_exhausted
E: Unable to allocate pthread_cond_t
 PASS - test_cond_resource_exhausted in 0.004 seconds
===================================================================
START - test_cond_resource_leak
 PASS - test_cond_resource_leak in 0.001 seconds
===================================================================
START - test_fnmatch
 PASS - test_fnmatch in 0.001 seconds
===================================================================
START - test_key_1toN_thread

Different threads set different values to same key:
thread 0: set value = -2147454144 and retrieved value = -2147454144
thread 1: set value = -2147453920 and retrieved value = -2147453920

 PASS - test_key_1toN_thread in 0.017 seconds
===================================================================
START - test_key_Nto1_thread

Single thread associates its value with different keys:
key 0: set value = -2147453720 and retrieved value = -2147453720
key 1: set value = -2147453720 and retrieved value = -2147453720

 PASS - test_key_Nto1_thread in 0.017 seconds
===================================================================
START - test_mqueue
 PASS - test_mqueue in 0.011 seconds
===================================================================
START - test_mutex_normal
E: k_mutex_unlock() failed: -16
mutex lock is taken
 PASS - test_mutex_normal in 0.105 seconds
===================================================================
START - test_mutex_recursive
recursive mutex lock is taken
 PASS - test_mutex_recursive in 0.003 seconds
===================================================================
START - test_mutex_resource_exhausted
E: Unable to allocate pthread_mutex_t
 PASS - test_mutex_resource_exhausted in 0.004 seconds
===================================================================
START - test_mutex_resource_leak
 PASS - test_mutex_resource_leak in 0.001 seconds
===================================================================
START - test_mutex_timedlock
Expecting timedlock with timeout of 200 ms to fail
E: Timeout locking mutex 0x80001218
Expecting timedlock with timeout of 200 ms to succeed after 100ms
 PASS - test_mutex_timedlock in 0.314 seconds
===================================================================
START - test_nanosleep_errors_errno
 PASS - test_nanosleep_errors_errno in 2.001 seconds
===================================================================
START - test_nanosleep_execution
 PASS - test_nanosleep_execution in 0.001 seconds
===================================================================
START - test_pthread_attr_stacksize
 PASS - test_pthread_attr_stacksize in 0.001 seconds
===================================================================
START - test_pthread_condattr
 PASS - test_pthread_condattr in 0.001 seconds
===================================================================
START - test_pthread_create_negative
E: Unable to allocate stack of size 0
E: pthread_attr_t is not initialized, has a NULL stack, or is of size 0
E: pthread_attr_t is not initialized, has a NULL stack, or is of size 0
 PASS - test_pthread_create_negative in 0.017 seconds
===================================================================
START - test_pthread_descriptor_leak
 PASS - test_pthread_descriptor_leak in 2.003 seconds
===================================================================
START - test_pthread_dynamic_stacks
 SKIP - test_pthread_dynamic_stacks in 0.001 seconds
===================================================================
START - test_pthread_equal
 PASS - test_pthread_equal in 0.001 seconds
===================================================================
START - test_pthread_errors_errno
E: Invalid pthread (ffffffff)
E: Invalid attr pointer
E: Invalid pthread_attr_t or sched_param
E: Invalid pthread (ffffffff)
E: Invalid pthread (ffffffff)
E: Invalid pthread (ffffffff)
E: Invalid pthread (ffffffff)
 PASS - test_pthread_errors_errno in 0.020 seconds
===================================================================
START - test_pthread_execution
E: Invalid pthread_attr_t or sched_param
E: NULL stack address
E: pthread_attr_t is not initialized, has a NULL stack, or is of size 0
Thread 0 starting with scheduling policy 1 & priority 15
E: Timeout locking mutex 0x80001218
Thread 1 starting with scheduling policy 1 & priority 15
E: Timeout locking mutex 0x80001218
Thread 2 starting with scheduling policy 1 & priority 15
E: Timeout locking mutex 0x80001218
Bounce test OK
Barrier test OK
 PASS - test_pthread_execution in 0.759 seconds
===================================================================
START - test_pthread_join_detached
E: Pthread 0x80000340 is not a joinable
 PASS - test_pthread_join_detached in 0.504 seconds
===================================================================
START - test_pthread_return_val
 PASS - test_pthread_return_val in 0.001 seconds
===================================================================
START - test_pthread_termination
E: Invalid pthread state -1
E: Invalid scheduler policy -1
E: Invalid piority -1 and / or policy 2
Thread 0 starting with a priority of 4
Cancelling thread 0
Thread 1 starting with a priority of 3
Cancelling thread 1
Thread 1 could not be cancelled
Thread 2 starting with a priority of 2
E: Pthread 0x80000640 cannot be detached
Cancelling thread 2
Thread 3 starting with a priority of 1
Cancelling thread 3
Thread 3 could not be cancelled
E: Pthread claims to be initialized (80000006)
E: Pthread 0x80000700 is not a joinable
E: Pthread attempted to join itself (8000000a)
E: Pthread claims to be initialized (80000006)
E: Pthread claims to be initialized (80000006)
 PASS - test_pthread_termination in 1.043 seconds
===================================================================
START - test_realtime
 PASS - test_realtime in 1.803 seconds
===================================================================
START - test_rw_lock

main acquire WR lock and 3 threads acquire RD lock
Thread 2 scheParent thread releasing WR lock
duling policParent thread acquiring WR lock again
Parent thread acquired WR lock again
y = 2 & priParent thread releasing WR lock again

3 threads acquire WR lock
Main thread acquiring RD lock
Main thread acquired RD lock
ority 3 starMain thread releasing RD lock
ted
Thread 2 got RD lock
Thread 1 schThread 2 releasing RD lock
Thread 2 acquiring WR lock
Thread 2 acquired WR lock
eduling polThread 2 releasing WR lock
icy = 2 & priority 2 started
Thread 1 got RD lock
Thread 0 scThread 1 releasing RD lock
Thread 1 acquiring WR lock
Thread 1 acquired WR lock
heduling polThread 1 releasing WR lock
icy = 2 & priority 1 started
Thread 0 got RD lock
Thread 0 releasing RD lock
Thread 0 acquiring WR lock
Thread 0 acquired WR lock
Thread 0 releasing WR lock
 PASS - test_rw_lock in 0.079 seconds
===================================================================
START - test_sched_policy
 PASS - test_sched_policy in 0.001 seconds
===================================================================
START - test_semaphore
 PASS - test_semaphore in 5.001 seconds
===================================================================
START - test_signal_addset
 PASS - test_signal_addset in 0.001 seconds
===================================================================
START - test_signal_addset_oor
 PASS - test_signal_addset_oor in 0.001 seconds
===================================================================
START - test_signal_delset
 PASS - test_signal_delset in 0.001 seconds
===================================================================
START - test_signal_delset_oor
 PASS - test_signal_delset_oor in 0.001 seconds
===================================================================
START - test_signal_emptyset
 PASS - test_signal_emptyset in 0.001 seconds
===================================================================
START - test_signal_fillset
 PASS - test_signal_fillset in 0.001 seconds
===================================================================
START - test_signal_ismember
 PASS - test_signal_ismember in 0.001 seconds
===================================================================
START - test_signal_ismember_oor
 PASS - test_signal_ismember_oor in 0.001 seconds
===================================================================
START - test_signal_strsignal
 PASS - test_signal_strsignal in 0.001 seconds
===================================================================
START - test_sleep
 PASS - test_sleep in 4.001 seconds
===================================================================
START - test_spin_descriptor_leak
 PASS - test_spin_descriptor_leak in 0.001 seconds
===================================================================
START - test_spin_init_destroy
 PASS - test_spin_init_destroy in 0.001 seconds
===================================================================
START - test_spin_lock_unlock
 PASS - test_spin_lock_unlock in 0.001 seconds
===================================================================
START - test_timer
POSIX timer test
Timer fires every 0 secs and  100000000 nsecs
Time remaining to fire 0 secs and  899000000 nsecs
Handler Signal value :20 for 1 times
Handler Signal value :20 for 2 times
Handler Signal value :20 for 3 times
Handler Signal value :20 for 4 times
Handler Signal value :20 for 5 times
Handler Signal value :20 for 6 times
Handler Signal value :20 for 7 times
Handler Signal value :20 for 8 times
Handler Signal value :20 for 9 times
Handler Signal value :20 for 10 times
Handler Signal value :20 for 11 times
Handler Signal value :20 for 12 times
 PASS - test_timer in 2.111 seconds
===================================================================
START - test_timer_overrun
 PASS - test_timer_overrun in 2.501 seconds
===================================================================
START - test_uname
 PASS - test_uname in 0.001 seconds
===================================================================
START - test_usleep
 PASS - test_usleep in 1.501 seconds
===================================================================
TESTSUITE posix_apis succeeded

------ TESTSUITE SUMMARY START ------

SUITE PASS - 100.00% [posix_apis]: pass = 48, fail = 0, skip = 1, total = 49 duration = 28.791 seconds
 - PASS - [posix_apis.test_barrier] duration = 0.001 seconds
 - PASS - [posix_apis.test_clock] duration = 1.005 seconds
 - PASS - [posix_apis.test_clock_nanosleep_errors_errno] duration = 0.001 seconds
 - PASS - [posix_apis.test_clock_nanosleep_execution] duration = 3.942 seconds
 - PASS - [posix_apis.test_cond_resource_exhausted] duration = 0.004 seconds
 - PASS - [posix_apis.test_cond_resource_leak] duration = 0.001 seconds
 - PASS - [posix_apis.test_fnmatch] duration = 0.001 seconds
 - PASS - [posix_apis.test_key_1toN_thread] duration = 0.017 seconds
 - PASS - [posix_apis.test_key_Nto1_thread] duration = 0.017 seconds
 - PASS - [posix_apis.test_mqueue] duration = 0.011 seconds
 - PASS - [posix_apis.test_mutex_normal] duration = 0.105 seconds
 - PASS - [posix_apis.test_mutex_recursive] duration = 0.003 seconds
 - PASS - [posix_apis.test_mutex_resource_exhausted] duration = 0.004 seconds
 - PASS - [posix_apis.test_mutex_resource_leak] duration = 0.001 seconds
 - PASS - [posix_apis.test_mutex_timedlock] duration = 0.314 seconds
 - PASS - [posix_apis.test_nanosleep_errors_errno] duration = 2.001 seconds
 - PASS - [posix_apis.test_nanosleep_execution] duration = 0.001 seconds
 - PASS - [posix_apis.test_pthread_attr_stacksize] duration = 0.001 seconds
 - PASS - [posix_apis.test_pthread_condattr] duration = 0.001 seconds
 - PASS - [posix_apis.test_pthread_create_negative] duration = 0.017 seconds
 - PASS - [posix_apis.test_pthread_descriptor_leak] duration = 2.003 seconds
 - SKIP - [posix_apis.test_pthread_dynamic_stacks] duration = 0.001 seconds
 - PASS - [posix_apis.test_pthread_equal] duration = 0.001 seconds
 - PASS - [posix_apis.test_pthread_errors_errno] duration = 0.020 seconds
 - PASS - [posix_apis.test_pthread_execution] duration = 0.759 seconds
 - PASS - [posix_apis.test_pthread_join_detached] duration = 0.504 seconds
 - PASS - [posix_apis.test_pthread_return_val] duration = 0.001 seconds
 - PASS - [posix_apis.test_pthread_termination] duration = 1.043 seconds
 - PASS - [posix_apis.test_realtime] duration = 1.803 seconds
 - PASS - [posix_apis.test_rw_lock] duration = 0.079 seconds
 - PASS - [posix_apis.test_sched_policy] duration = 0.001 seconds
 - PASS - [posix_apis.test_semaphore] duration = 5.001 seconds
 - PASS - [posix_apis.test_signal_addset] duration = 0.001 seconds
 - PASS - [posix_apis.test_signal_addset_oor] duration = 0.001 seconds
 - PASS - [posix_apis.test_signal_delset] duration = 0.001 seconds
 - PASS - [posix_apis.test_signal_delset_oor] duration = 0.001 seconds
 - PASS - [posix_apis.test_signal_emptyset] duration = 0.001 seconds
 - PASS - [posix_apis.test_signal_fillset] duration = 0.001 seconds
 - PASS - [posix_apis.test_signal_ismember] duration = 0.001 seconds
 - PASS - [posix_apis.test_signal_ismember_oor] duration = 0.001 seconds
 - PASS - [posix_apis.test_signal_strsignal] duration = 0.001 seconds
 - PASS - [posix_apis.test_sleep] duration = 4.001 seconds
 - PASS - [posix_apis.test_spin_descriptor_leak] duration = 0.001 seconds
 - PASS - [posix_apis.test_spin_init_destroy] duration = 0.001 seconds
 - PASS - [posix_apis.test_spin_lock_unlock] duration = 0.001 seconds
 - PASS - [posix_apis.test_timer] duration = 2.111 seconds
 - PASS - [posix_apis.test_timer_overrun] duration = 2.501 seconds
 - PASS - [posix_apis.test_uname] duration = 0.001 seconds
 - PASS - [posix_apis.test_usleep] duration = 1.501 seconds

------ TESTSUITE SUMMARY END ------

===================================================================
PROJECT EXECUTION SUCCESSFUL
'''

cfriedt commented on tests/posix/common/src/nanosleep.c:
```diff
@@ -143,7 +143,7 @@ static void common_lower_bound_check(int selection, clockid_t clock_id, int flag
 	zassert_equal(rem.tv_nsec, 0, "actual: %d expected: %d", rem.tv_nsec, 0);
 
-	actual_ns = k_cyc_to_ns_ceil64((now - then * selection));
+	actual_ns = k_cyc_to_ns_ceil64((now + then * selection));
```
Actually, this is just measuring a duration, so it should be independent of selection.

This looks like it was a regression introduced in the `clock_nanosleep()` PR
```suggestion
	actual_ns = k_cyc_to_ns_ceil64(now - then);
```

hakehuang replied:
@cfriedt ,  it is related to the seletction, as last commit, the 

```
static inline int select_nanosleep(int selection, clockid_t clock_id, int flags,
				   const struct timespec *rqtp, struct timespec *rmtp)
{
	if (selection == SELECT_NANOSLEEP) {
		return nanosleep(rqtp, rmtp);
	}
	return clock_nanosleep(clock_id, flags, rqtp, rmtp);
}
```
so this code is reused for different `flags`,  and when flags == 0, we need minus the then time(uptime), but with flags==4(TIMER_ABSTIME), we do not need do this.

cfriedt replied:
Yeah, I can see now why the author decided to change the test this way. Personally, I would have preferred if the absolute time was a separate check in each case, but that would have duplicated several checks throughout the suite.

Can you adjust this change to better document the expectation here?
```diff
diff --git a/tests/posix/common/src/nanosleep.c b/tests/posix/common/src/nanosleep.c
index 5e4d3bcf88..a21309e1c8 100644
--- a/tests/posix/common/src/nanosleep.c
+++ b/tests/posix/common/src/nanosleep.c
@@ -11,8 +11,10 @@
 #include <zephyr/sys_clock.h>
 #include <zephyr/ztest.h>
 
-#define SELECT_NANOSLEEP       1
+/** @brief Test @ref clock_nanosleep */
 #define SELECT_CLOCK_NANOSLEEP 0
+/** @brief Test @ref nanosleep */
+#define SELECT_NANOSLEEP       1
 
 static inline int select_nanosleep(int selection, clockid_t clock_id, int flags,
                                   const struct timespec *rqtp, struct timespec *rmtp)
@@ -120,6 +122,22 @@ ZTEST(posix_apis, test_clock_nanosleep_errors_errno)
        zassert_equal(rem.tv_nsec, 0, "actual: %d expected: %d", rem.tv_nsec, 0);
 }
 
+/**
+ * @brief Check that a call to nanosleep has yielded executiuon for some minimum time.
+ *
+ * Check that the actual time slept is >= the total time specified by @p s (in seconds) and
+ * @p ns (in nanoseconds).
+ *
+ * @note The time specified by @p s and @p ns is assumed to be absolute (i.e. a time-point)
+ * when @p selection is set to @ref SELECT_CLOCK_NANOSLEEP. The time is assumed to be relative
+ * when @p selection is set to @ref SELECT_NANOSLEEP.
+ *
+ * @param selection Either @ref SELECT_CLOCK_NANOSLEEP or @ref SELECT_NANOSLEEP
+ * @param clock_id The clock to test (e.g. @ref CLOCK_MONOTONIC or @ref CLOCK_REALTIME)
+ * @param flags Flags to pass to @ref clock_nanosleep
+ * @param s Partial lower bound for yielded time (in seconds)
+ * @param ns Partial lower bound for yielded time (in nanoseconds)
+ */
 static void common_lower_bound_check(int selection, clockid_t clock_id, int flags, const uint32_t s,
                                     uint32_t ns)
 {
@@ -143,7 +161,18 @@ static void common_lower_bound_check(int selection, clockid_t clock_id, int flag
        zassert_equal(rem.tv_sec, 0, "actual: %d expected: %d", rem.tv_sec, 0);
        zassert_equal(rem.tv_nsec, 0, "actual: %d expected: %d", rem.tv_nsec, 0);
 
-       actual_ns = k_cyc_to_ns_ceil64((now - then * selection));
+       switch(selection) {
+               case SELECT_NANOSLEEP:
+                       /* exp_ns and actual_ns are relative (i.e. durations) */
+                       actual_ns = k_cyc_to_ns_ceil64(now - then);
+                       break;
+               case SELECT_CLOCK_NANOSLEEP:
+                       /* exp_ns and actual_ns are absolute (i.e. time-points) */
+                       actual_ns = k_cyc_to_ns_ceil64(now);
+                       break;
+               default:
+                       break;
+       }
 
        exp_ns = (uint64_t)s * NSEC_PER_SEC + ns;
        /* round up to the nearest microsecond for k_busy_wait() */
```

cfriedt replied:
I guess maybe an assert in the default selection would be good..

hakehuang replied:
updated

cfriedt reviewed (commented):
Minor correction. Otherwise LGTM

hakehuang commented on tests/posix/common/src/nanosleep.c:
```diff
@@ -126,15 +126,23 @@ static void common_lower_bound_check(int selection, clockid_t clock_id, int flag
 	int r;
 	uint64_t actual_ns;
 	uint64_t exp_ns;
-	uint32_t now;
-	uint32_t then;
+	uint64_t now;
+	uint64_t then;
 	struct timespec rem = {0, 0};
 	struct timespec req = {s, ns};
 
 	errno = 0;
+#if CONFIG_TIMER_HAS_64BIT_CYCLE_COUNTER
+	then = k_cycle_get_64();
+#else
 	then = k_cycle_get_32();
+#endif
 	r = select_nanosleep(selection, clock_id, flags, &req, &rem);
+#if CONFIG_TIMER_HAS_64BIT_CYCLE_COUNTER
+	now = k_cycle_get_64();
+#else
 	now = k_cycle_get_32();
+#endif
```
@cfriedt , the ksleep is a usleep, which ROUND_UP the ns sleep time, so there will be some accuracy problem in our platform when the sleep time is very short, so I change it to k_cycle_get_64, which will get more pricise clock, however, Nordic simulation platform does not support k_cycle_get_64 so I add this code, this is to resolve the problem introduced by this `+then` fixes.

cfriedt replied:
Great - can you love it to a separate commit please? I'll add some additional change requests for that code now so that you don't need to respin later.

hakehuang replied:
done

cfriedt commented on tests/posix/common/src/nanosleep.c:
```diff
@@ -143,7 +167,18 @@ static void common_lower_bound_check(int selection, clockid_t clock_id, int flag
+		/* exp_ns and actual_ns are absolute (i.e. time-points) */
+		actual_ns = k_cyc_to_ns_ceil64(now);
+		break;
+	default:
```
Please add an assertion here to fail in the default case.

hakehuang replied:
I see, annd add a unreachable assert in default 
'''
zassert_unreachable();
'''

cfriedt commented on tests/posix/common/src/nanosleep.c:
```diff
@@ -120,21 +120,45 @@ ZTEST(posix_apis, test_clock_nanosleep_errors_errno)
 	struct timespec req = {s, ns};
 
 	errno = 0;
+#if CONFIG_TIMER_HAS_64BIT_CYCLE_COUNTER
+	then = k_cycle_get_64();
+#else
 	then = k_cycle_get_32();
+#endif
 	r = select_nanosleep(selection, clock_id, flags, &req, &rem);
+#if CONFIG_TIMER_HAS_64BIT_CYCLE_COUNTER
+	now = k_cycle_get_64();
+#else
 	now = k_cycle_get_32();
+#endif
```
Please add 64-bit change to a separate commit.

This can go up top (let's avoid using `#ifdef` everywhere).
```cpp
static inline uint64_t cycle_get_64(void)
{
    if (IS_ENABLED(CONFIG_TIMER_HAS_64BIT_CYCLE_COUNTER) {
        return k_cycle_get_64();
    } else {
        return k_cycle_get_32();
    }
}
```

```suggestion
	then = cycle_get_64();
 	r = select_nanosleep(selection, clock_id, flags, &req, &rem);
	now = cycle_get_64();
```

hakehuang replied:
updated. Thanks for this recommendation. @cfriedt

hakehuang committed:
test: posix: nanosleep: use k_cycle_64 as possible

as posix nanosleep is actually call us, so we need use a
higher clock to compare the result. try to use k_cycle_64
as possible.

Co-authored-by: Christopher Friedt <cfriedt@meta.com>
Signed-off-by: Hake Huang <hake.huang@oss.nxp.com>
